// @progress/kendo-react-layout/dist/npm/panelbar/services/navigationService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var kendo_react_common_1 = require("@progress/kendo-react-common");
/**
 * @hidden
 */
var NavigationAction;
(function (NavigationAction) {
    NavigationAction[NavigationAction["Toggle"] = 0] = "Toggle";
    NavigationAction[NavigationAction["Next"] = 1] = "Next";
    NavigationAction[NavigationAction["Previous"] = 2] = "Previous";
})(NavigationAction = exports.NavigationAction || (exports.NavigationAction = {}));
/**
 * @hidden
 */
var NavigationService = /** @class */ (function () {
    function NavigationService(onNavigate) {
        this.onNavigate = onNavigate;
    }
    NavigationService.prototype.navigate = function (event, rtl) {
        var keyCode = event.keyCode;
        var action;
        switch (keyCode) {
            case kendo_react_common_1.Keys.left:
                action = !rtl ? NavigationAction.Previous : NavigationAction.Next;
                break;
            case kendo_react_common_1.Keys.up:
                action = NavigationAction.Previous;
                break;
            case kendo_react_common_1.Keys.right:
                action = !rtl ? NavigationAction.Next : NavigationAction.Previous;
                break;
            case kendo_react_common_1.Keys.down:
                action = NavigationAction.Next;
                break;
            case kendo_react_common_1.Keys.space:
                action = NavigationAction.Toggle;
                break;
            case kendo_react_common_1.Keys.enter:
                action = NavigationAction.Toggle;
                break;
            default:
                action = null;
                break;
        }
        if (action !== null) {
            event.preventDefault();
            this.onNavigate(action);
        }
    };
    return NavigationService;
}());
exports.default = NavigationService;
