// @progress/kendo-react-layout/dist/npm/card/interfaces/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var avatarShape;
(function (avatarShape) {
    avatarShape["CIRCLE"] = "circle";
    avatarShape["SQUARE"] = "square";
    avatarShape["ROUNDED"] = "rounded";
})(avatarShape = exports.avatarShape || (exports.avatarShape = {}));
var avatarType;
(function (avatarType) {
    avatarType["INITIALS"] = "initials";
    avatarType["IMAGE"] = "image";
    avatarType["ICON"] = "icon";
})(avatarType = exports.avatarType || (exports.avatarType = {}));
var cardOrientation;
(function (cardOrientation) {
    cardOrientation["HORIZONTAL"] = "horizontal";
    cardOrientation["VERTICAL"] = "vertical";
})(cardOrientation = exports.cardOrientation || (exports.cardOrientation = {}));
var cardType;
(function (cardType) {
    cardType["DEFAULT"] = "default";
    cardType["primary"] = "primary";
    cardType["INFO"] = "info";
    cardType["SUCCESS"] = "success";
    cardType["WARNING"] = "warning";
    cardType["ERROR"] = "error";
})(cardType = exports.cardType || (exports.cardType = {}));
var cardActionsLayout;
(function (cardActionsLayout) {
    cardActionsLayout["START"] = "start";
    cardActionsLayout["CENTER"] = "center";
    cardActionsLayout["END"] = "end";
    cardActionsLayout["STRETCHED"] = "stretched";
})(cardActionsLayout = exports.cardActionsLayout || (exports.cardActionsLayout = {}));
