// @progress/kendo-react-layout/dist/npm/menu/utils/itemsIdsUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @hidden
 */
exports.EMPTY_ID = '';
/**
 * @hidden
 */
exports.ZERO_LEVEL_ZERO_ITEM_ID = '0';
/**
 * @hidden
 */
exports.SEPARATOR = '_';
/**
 * @hidden
 */
function getItemById(itemId, items) {
    if (isIdZeroLevel(itemId)) {
        return items[Number(itemId)];
    }
    else {
        var rootParentItem = items[Number(getRootParentId(itemId))];
        return rootParentItem.items ?
            getItemById(getIdWithoutRootParentId(itemId), rootParentItem.items) : undefined;
    }
}
exports.getItemById = getItemById;
/**
 * @hidden
 */
function getRootParentId(itemId) {
    return isIdEmptyOrZeroLevel(itemId) ? itemId : itemId.split(exports.SEPARATOR)[0];
}
exports.getRootParentId = getRootParentId;
/**
 * @hidden
 */
function getIdWithoutRootParentId(itemId) {
    if (isIdEmptyOrZeroLevel(itemId)) {
        return itemId;
    }
    else {
        var firstSeparatorIndex = itemId.indexOf(exports.SEPARATOR);
        return itemId.substring(firstSeparatorIndex + 1);
    }
}
exports.getIdWithoutRootParentId = getIdWithoutRootParentId;
/**
 * @hidden
 */
function getFirstChildId(itemId) {
    return createId('0', itemId);
}
exports.getFirstChildId = getFirstChildId;
/**
 * @hidden
 */
function shouldOpenItem(itemId, lastItemIdToBeOpened) {
    if (lastItemIdToBeOpened.indexOf(itemId) === 0) {
        return lastItemIdToBeOpened.length === itemId.length ||
            lastItemIdToBeOpened.charAt(itemId.length) === exports.SEPARATOR;
    }
    else {
        return false;
    }
}
exports.shouldOpenItem = shouldOpenItem;
/**
 * @hidden
 */
function createId(childId, parentId) {
    return parentId ? parentId + exports.SEPARATOR + childId : childId;
}
exports.createId = createId;
/**
 * @hidden
 */
function getDirectParentId(itemId) {
    var lastSeparatorIndex = itemId.lastIndexOf(exports.SEPARATOR);
    return lastSeparatorIndex < 0 ? exports.EMPTY_ID : itemId.substring(0, lastSeparatorIndex);
}
exports.getDirectParentId = getDirectParentId;
/**
 * @hidden
 */
function isIdEmptyOrZeroLevel(itemId) {
    return itemId === exports.EMPTY_ID || itemId.indexOf(exports.SEPARATOR) < 0;
}
exports.isIdEmptyOrZeroLevel = isIdEmptyOrZeroLevel;
/**
 * @hidden
 */
function isIdZeroLevel(itemId) {
    return itemId !== exports.EMPTY_ID && itemId.indexOf(exports.SEPARATOR) < 0;
}
exports.isIdZeroLevel = isIdZeroLevel;
/**
 * @hidden
 */
function isIdFirstLevel(itemId) {
    return getSeparatorOccurances(itemId) === 1;
}
exports.isIdFirstLevel = isIdFirstLevel;
/**
 * @hidden
 */
function isFirstItemFromSiblings(itemId) {
    return getShortId(itemId) === exports.ZERO_LEVEL_ZERO_ITEM_ID;
}
exports.isFirstItemFromSiblings = isFirstItemFromSiblings;
/**
 * @hidden
 */
function getShortId(itemId) {
    var lastSeparatorIndex = itemId.lastIndexOf(exports.SEPARATOR);
    return lastSeparatorIndex < 0 ? itemId : itemId.substring(lastSeparatorIndex + 1);
}
exports.getShortId = getShortId;
/**
 * @hidden
 */
function getDirectSiblingIdForLevelZero(next, shortItemId, siblingsCount) {
    if (!isIdZeroLevel(shortItemId)) {
        return shortItemId;
    }
    return next ?
        Number(shortItemId) < siblingsCount - 1 ? (Number(shortItemId) + 1).toString() : '0' :
        Number(shortItemId) > 0 ? (Number(shortItemId) - 1).toString() : (siblingsCount - 1).toString();
}
exports.getDirectSiblingIdForLevelZero = getDirectSiblingIdForLevelZero;
function getSeparatorOccurances(itemId) {
    return itemId.split(exports.SEPARATOR).length - 1;
}
