// @progress/kendo-react-layout/dist/npm/menu/components/MenuItemArrow.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var misc_1 = require("../utils/misc");
/**
 * @hidden
 */
exports.downArrowClass = 'k-i-arrow-60-down';
/**
 * @hidden
 */
exports.rightArrowClass = 'k-i-arrow-60-right';
/**
 * @hidden
 */
exports.leftArrowClass = 'k-i-arrow-60-left';
var MenuItemArrow = /** @class */ (function (_super) {
    __extends(MenuItemArrow, _super);
    function MenuItemArrow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @hidden
     */
    MenuItemArrow.prototype.render = function () {
        return React.createElement("span", { className: this.getArrowClassName(), role: "presentation" });
    };
    MenuItemArrow.prototype.getArrowClassName = function () {
        var _a;
        var childrenPosition = misc_1.getChildrenPosition(this.props.itemId, this.props.verticalMenu === true, this.props.dir === 'rtl');
        return kendo_react_common_1.classNames('k-icon', 'k-menu-expand-arrow', (_a = {},
            _a[exports.downArrowClass] = childrenPosition === 'downward',
            _a[exports.rightArrowClass] = childrenPosition === 'rightward',
            _a[exports.leftArrowClass] = childrenPosition === 'leftward',
            _a));
    };
    /**
     * @hidden
     */
    MenuItemArrow.propTypes = {
        itemId: PropTypes.string,
        dir: PropTypes.string,
        verticalMenu: PropTypes.bool
    };
    return MenuItemArrow;
}(React.Component));
exports.default = MenuItemArrow;
